experimental/sev_guest/src/instructions.rs

```diff
@@ -0,0 +1,193 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Rust instruction wrappers for managing page state and interacting with the hypervisor.
+
+use bitflags::bitflags;
+use core::arch::asm;
+use strum::FromRepr;
+
+/// Whether a page is in the validated state or not.
+#[derive(Debug, FromRepr)]
+#[repr(u32)]
+pub enum Validation {
+    /// The page is not validated.
+    Unvalidated = 0,
+    /// The page is validated.
+    Validated = 1,
+}
+
+/// The size of a memory page.
+#[derive(Debug, FromRepr)]
+#[repr(u32)]
+pub enum PageSize {
+    /// The page is a 4KiB page.
+    Page4KiB = 0,
+    /// The page is a 2MiB page.
+    Page2MiB = 1,
+}
+
+/// The potential errors when calling the PVALIDATE or RMPADJUST instructions.
+#[derive(Debug, FromRepr)]
+#[repr(u32)]
+pub enum InstructionError {
+    /// The input parameters were invalid.
+    FailInput = 1,
+    /// Insufficient permissions.
+    FailPermission = 2,
+    /// The page size does not match the page size entry in the RMP.
+    FailSizeMismatch = 6,
+    /// The page validation status was not updated. This value is software defined and will not be
+    /// returned by the hardware instruction.
+    ValdationStatusNotUpdated = 255,
+}
+
+/// Marks a page as validated or unvalidated in the RMP.
+///
+/// See the PVALIDATE instruction in <https://www.amd.com/system/files/TechDocs/24594.pdf> for more details.
+#[inline]
+pub fn pvalidate(
+    page_addr: u64,
```

Comment:
a) Is this a physical or virtual address?
b) I think `usize` would be a better semantic match than raw `u64`.

Reply:
These are guest physical addresses. Updated the parameter name to clarify. Changed to usize, although these instructions are only available in 64 bit mode, so u64 and usize is equivalent.